datafusion/physical-plan/src/joins/cross_join.rs

```diff
@@ -374,64 +346,130 @@ impl Stream for CrossJoinStream {
-                    let result =
-                        build_batch(self.left_index, &batch, left_data, &self.schema);
+            // If it is the first poll or the right batch is fully processed:
+            _ => match ready!(self.right.poll_next_unpin(cx)) {
+                None => Poll::Ready(None),
+                Some(Ok(right_batch)) => {
+                    // New batch arrives, reset the indices.
+                    self.left_batch_index = 0;
+                    self.right_row_index = 0;
+                    let result_batch = build_batch(
+                        &mut self.left_batch_index,
+                        &mut self.right_row_index,
+                        &mut self.join_metrics,
+                        left_data,
+                        &right_batch,
+                        schema,
+                    )?;
+                    // Update the metrics.
                     self.join_metrics.input_batches.add(1);
-                    self.join_metrics.input_rows.add(batch.num_rows());
-                    if let Ok(ref batch) = result {
-                        join_timer.done();
-                        self.join_metrics.output_batches.add(1);
-                        self.join_metrics.output_rows.add(batch.num_rows());
-                    }
-                    self.left_index = 1;
-
-                    let mut right_batch = self.right_batch.lock();
-                    *right_batch = Some(batch);
-
-                    Some(result)
+                    self.join_metrics.input_rows.add(
+                        self.curr_probe_batch.as_ref().map_or(0, |rb| rb.num_rows()),
+                    );
+                    // Store the new batch into the state.
+                    self.curr_probe_batch = Some(right_batch);
+                    Poll::Ready(Some(Ok(result_batch)))
                 }
-                other => other,
-            })
+                Some(Err(err)) => Poll::Ready(Some(Err(err))),
+            },
+        }
+    }
+}
+
+/// This function constructs a new `RecordBatch` by joining the left and right batches
+/// based on the current indices. It also updates the metrics.
+///
+/// # Arguments
+/// * `left_batch_index` - Index of the current left batch being processed.
+/// * `right_row_index` - Index of the current row in the right batch to be joined.
+/// * `join_metrics` - Metrics container to track performance of the join operation.
+/// * `left_data` - Array of `RecordBatch`es from the left side to be joined.
+/// * `probe_batch` - The current `RecordBatch` from the right side to be joined.
+/// * `schema` - `Arc<Schema>` describing the schema of the resulting `RecordBatch`.
+fn build_batch(
+    left_batch_index: &mut usize,
+    right_row_index: &mut usize,
+    join_metrics: &mut BuildProbeJoinMetrics,
+    left_data: &[RecordBatch],
+    probe_batch: &RecordBatch,
+    schema: Arc<Schema>,
+) -> Result<RecordBatch> {
+    if probe_batch.num_rows() != 0 {
```

Comment:
This is one of the reasons why we use state machines in joins. You do not have a reliable re-try mechanism, thus you become obligated to send a empty record batch, which is not a good way to handle this since it creates overhead. If a probe batch is empty, retry. use loop and state machine mechanisms for a better understanding of the algorithm.